tactics/rewrite.ml

```diff
@@ -1600,7 +1603,12 @@ let cl_rewrite_clause_newtac ?abs ?origsigma ~progress strat clause =
       beta <*> Proofview.shelve_unifiable
     with
     | PretypeError (env, evd, (UnsatisfiableConstraints _ as e)) ->
-      raise (RewriteFailure (env, evd, e))
+      if do_rewrite_output_constraints () then
+        let fold ev _ accu = ev :: accu in
+        let gls = List.rev (Evd.fold_undefined fold evd []) in
+        let gls = List.map (fun gl -> Proofview.goal_with_state gl state) gls in
+         Proofview.Unsafe.tclEVARS evd <*> Proofview.Unsafe.tclNEWGOALS gls
```

Comment:
Note that this is a bit fragile w.r.t. goal ordering. I assume it's fine because this is basically a debug flag, but if somebody starts depending on that we should either have a more principled algorithm or simply state outright that users shouldn't depend on the order.

Reply:
That's a very good point. People won't shouldn't "rely" on it a priori as it is a debug flag, but it would be better to give the constraints in the same order as typeclass resolution sees them, I'll improve it.

Reply:
we could randomize the order